### PR TITLE
 [Bigtable] Documented defaults timeout values in the EnhancedBigtableStubSettings

### LINKED ISSUE #2992: bigtable: update javadocs for settings to describe the defaults
This is extracted from: https://github.com/GoogleCloudPlatform/google-cloud-java/pull/2986/files/5ce135c8e7f55fd510382574357b14a775f5619b#r171952810

The javadocs should document the default timeouts

Please assign this to me

### PR DESCRIPTION
**Pull request by rahulKQL**
Fixes #2992

Updated default timeouts for all the operations in `EnhnaceDBigtableStubSettings` as `BigtableClientSettings` settings were deprecated.

### CODE REVIEW

**Comment by rahulKQL** [comment]
@igorbernstein2 Please have a look at this change. 
Please let me know if the format can be improved for timeout listing.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 1]
```diff
@@ -55,6 +55,24 @@
...
  *   <li>Credentials are acquired automatically through Application Default Credentials.
  *   <li>Retries are configured for idempotent methods but not for non-idempotent methods.
+ *   <li>These are the default timeouts used Bigtable operations.
+ *       <ul>
+ *         For non-streaming operations i.e. {@link #readRowSettings()}, {@link
+ *         #sampleRowKeysSettings()}, {@link #checkAndMutateRowSettings()} {@link
+ *         #mutateRowSettings()}, {@link #bulkMutateRowsSettings()} {@link
+ *         #readModifyWriteRowSettings()}:
+ *         <li>Default timeout for {@link RetrySettings#getInitialRpcTimeout()} is 20_000ms.
+ *         <li>Default timeout for {@link RetrySettings#getMaxAttempts()} is 20_000ms.
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: timeout for max attempts as a duration?

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 2]
```diff
@@ -55,6 +55,24 @@
  *   <li>The default service address (bigtable.googleapis.com) and default port (443) are used.
  *   <li>Credentials are acquired automatically through Application Default Credentials.
  *   <li>Retries are configured for idempotent methods but not for non-idempotent methods.
+ *   <li>These are the default timeouts used Bigtable operations.
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: This needs a lot of work and should probably be in its own paragraph and/or in each method doc:
- A distinction should be drawn between idempotent methods (ReadRow(s), MutateRow(s), SampleRowKeys) and non-idempotent methods (CheckAndMutateRows, ReadModiifyWrite)
- The distinction for streaming vs non-streaming should be kept. But there needs to be some explanation about how the settings apply. For example rpc timeout for non-streaming is the deadline for each call, but for streaming the rpc timeout is the timeout between each row. 
- There should be docs explaining what rpc timeout is vs total timeout (ie. attempt timeout vs operation timeout)
- The values should be human readable, ie instead of 20_000ms, it should be 20 seconds
- There should be docs about exponential delay defaults 
- There should be an explanation of what idleTimeout is 
- The formatting is currently invalid: ```<ul>TEXT<li>first element text``` is not valid markup

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java (hunk 1 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: My apologies, It was a typo from my end it should be `getMaxRpcTimeout`.

Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: I would collapse the rpc timeout min/max/multiplier settings into a single constant rpc timeout value in the docs

**Comment by igorbernstein** [comment]
@kennethlowery8190, please remove the approval. This pr needs more work and I don’t want it accidentally merged

**Comment by rahulKQL** [comment]
@igorbernstein2  Please have another round of review. 

I have added below points in the last commit.
- Listed Description for streaming and non-streaming operation.
- Tried explaining RPC vs Total timeout
- Tried explaining retryDelayMultiplier.
- Added default values on each setting operation's Javadoc.
- Added bulkMutateRowsSettings description.

**Comment by sduskis** [comment]
@igorbernstein2 or @kolea2, can you PTAL

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 3]
```diff
@@ -55,6 +55,13 @@
  *   <li>The default service address (bigtable.googleapis.com) and default port (443) are used.
  *   <li>Credentials are acquired automatically through Application Default Credentials.
  *   <li>Retries are configured for idempotent methods but not for non-idempotent methods.
+ *   <li>This settings are mixed for streaming and non-streaming operation. In case of streaming
+ *       operations, RPC timeout applies to each row and for the non-streaming operations, RPC
+ *       timeout considered as deadline for every call.
+ *   <li>Each operation setting accepts RPC and total time out. Here, RPC timeouts are for each
+ *       attempt to get the result whereas total timeout applies to the operation timeout.
+ *   <li>RetryDelayMultiplier controls the change in retry delay. After initial attempt, subsequent
+ *       attempts are calculated by multiplying the retry delay of the previous call.
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: This feels out of place. The first  3 bullet points describe the values of the settings, while new 3 points are describing concepts. Also, what are you referring to by `this settings`? Also, it's weird that the explanation for RetryDelayMultiplier is 3 bullet points away from the Retries bullet point.

I think it would be better to link users to read the docs for RetrySettings for a description of the concepts and focus on the values in these docs. (see below)

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 4]
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
...
 
-  /** Returns the object with the settings used for calls to ReadRows. */
+  /**
+   * Returns the object with the settings used for calls to ReadRows.
+   *
+   * <p>This is idempotent and streaming operation.
+   *
+   * <p>The idle timeout is configured via {@code ServerStreamingCallSettings#getIdleTimeout}. This
+   * timer will terminate any stream that has not has seen any demand or activity in the configured
+   * interval. To turn off idle checks, set the idleTimeout to {@link Duration#ZERO}.
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Please document the default value here (5 minutes)

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 5]
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
...
+   *
+   * <p>Retries will be attempted if RPC failed with {@link Code#DEADLINE_EXCEEDED} and {@link
+   * Code#UNAVAILABLE} failure code.
+   *
+   * <p>RetryDelay starts with 100ms or 0.1 seconds, which multiplies with 1.3 times on every
+   * reattempt till a max retry timeout value of 60 seconds.
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Please drop `0.1 seconds`. Also, this isnt a timeout. Maybe reword to:

The RetryDelay between failed attempts starts with 100ms and exponentially increases by a factor of 1.3 until 60 seconds.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 6]
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
...
+   *
+   * <p>RPC timeout is 20 seconds.
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: I think you should combine the 2 timeouts docs and call out that the RPC timeout means wait timeout:

The default read timeout for each row in a response stream is 20 seconds. And the timeout to read the entire stream is 1 hour.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 7]
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
...
+   *
+   * <p>Total timeout for this operation is 600 seconds or 10 mins.
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Similar to the streaming example I think you should combine the timeouts:

The default timeout for  each attempt is 20 seconds and the timeout for the entire operation across all of the attempts is 10 mins.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 8]
```diff
@@ -185,22 +192,92 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
         .setJwtEnabledScopes(JWT_ENABLED_SCOPES);
   }
 
-  /** Returns the object with the settings used for calls to ReadRows. */
+  /**
+   * Returns the object with the settings used for calls to ReadRows.
+   *
+   * <p>This is idempotent and streaming operation.
+   *
+   * <p>The idle timeout is configured via {@code ServerStreamingCallSettings#getIdleTimeout}. This
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Maybe nest these in a list? Also as opposed to duplicating the docs in RetrySettings and ServerStreamingSettings, just link out to them. Also I think it would better to link directly to the setters

Maybe something this?

```
<p>Default settings:
<ul>
  <li>{@link ServerStreamingCallSettings.Builder#setIdleTimeout Default idle timeout} is set to 1 hour
  <li>{@link ServerStreamingCallSettings.Builder#setRetryableCodes Error codes} are: {@link Code#DEADLINE_EXCEEDED} and {@link Code#UNAVAILABLE}
<li>RetryDelay {@link RetrySettings.Builder#setInitialRetryDelay starts} at 100ms and {@link RetrySettings.Builder#setRetryDelayMultiplier increases exponentially} by 1.3 until a maximum of ....


```

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java (hunk 4 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: I have rephrased this line with default time.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java (hunk 5 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Thanks for the detailed suggestion, I have updated these settings with the same.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java (hunk 3 above)
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: I think with your new suggested explanation, User will get the idea about these Retries and RPC timeouts so I have removed these explanations from the Class level Javadoc. 
Please let me know if we want to add any details here.

**Comment by rahulKQL** [comment]
@igorbernstein2 I tried to address feedback comments, Please take a look.

**Comment by igorbernstein2** [comment]
Just a couple minor nits, but LGTM!

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 9]
```diff
@@ -185,22 +185,100 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
...
+   * <p>This is idempotent and streaming operation.
+   *
+   * <p>Default settings:
+   *
+   * <ul>
+   *   <li>{@link ServerStreamingCallSettings.Builder#setIdleTimeout Default idle timeout} is set to
+   *       5 mins.
+   *   <li>Retry {@link ServerStreamingCallSettings.Builder#setRetryableCodes error codes} are:
+   *       {@link Code#DEADLINE_EXCEEDED} and {@link Code#UNAVAILABLE}.
+   *   <li>RetryDelay between failed attempt {@link RetrySettings.Builder#setInitialRetryDelay
```
[the remark this answers is not in the document]
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Thanks for the review, Have fixed this typo.

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java [hunk 10]
```diff
@@ -185,22 +185,100 @@ public static TransportChannelProvider defaultTransportChannelProvider() {
...
+   *
+   * <ul>
+   *   <li>Retry {@link UnaryCallSettings.Builder#setRetryableCodes error codes} are: {@link
+   *       Code#DEADLINE_EXCEEDED} and {@link Code#UNAVAILABLE}.
+   *   <li>RetryDelay between failed attempt {@link RetrySettings.Builder#setInitialRetryDelay
+   *       starts} at 100ms and {@link RetrySettings.Builder#setRetryDelayMultiplier increases
+   *       exponentially} by a factor of 1.3 until a {@link RetrySettings.Builder#setMaxRetryDelay
+   *       maximum of} 60 seconds.
+   *   <li>The default timeout for {@link RetrySettings.Builder#setMaxRpcTimeout each attempt} is 20
+   *       seconds and the timeout for {@link RetrySettings.Builder#setTotalTimeout entire
```
[the remark this answers is not in the document]
Comment (rahulKQL) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java]: Addressed this as well.

**Comment by rahulKQL** [comment]
@igorbernstein2 Have addressed your feedback, Please have another look.

**Comment by igorbernstein2** [comment]
Look like a spanner test is failing
